### DISCUSSION TITLE
Tips for running GMD node on Raspberry Pi Zero

### DISCUSSION THREAD

**Discussion by georgeasiminei** [CoopNetwork/GMD-blockchain#27] [Show and tell]
1. For OS I used Raspberry Pi OS Lite (32 bit) based on Debian Bullseye
![image](https://user-images.githubusercontent.com/4907747/145266136-475414f1-8de8-4fcf-b928-845e8a932cb1.png)

2. As the old Raspberry Pi Zero is based on ARMv6 architecture, only Java 8 will work. To install Java 8:
```
sudo apt update
sudo apt install -y openjdk-8-jdk-headless
```
2. [Optional] As Pi Zero has only 512MB of RAM, you will need to add some memory disk swap. It is recommended to have somehow fast storage (sdcard or external usb drive). You can add an additional 1GB of swap with following commands:

  - `sudo fallocate -l 1G /swapfile`

  - If the fallocate utility is not present on your system, or you get an error message saying fallocate failed: Operation not supported, use the following command to create the swap file: sudo dd if=/dev/zero of=/swapfile bs=1024 count=2097152
``` 
sudo chmod 600 /swapfile
sudo mkswap /swapfile
sudo swapon /swapfile
```
  - `sudo vi /etc/fstab` and add this line: `/swapfile swap swap defaults 0 0`
  - to verify `sudo swapon --show`
3. Compile the code `./compile.sh --skip-desktop`
4. The CPU usage of the node is very low in normal operation so Pi Zero should be enough. 
    - Exception to this rule:  A huge CPU load is needed when starting the node for the first time when the database is created and all genesis accounts are loaded. Pi Zero cannot handle this load (either it would take lots of hours to start, either it will crash).
    -  The solution to overcome this issue is to start a GMD node on another machine with a better CPU and 1GB of RAM (OS is not important - wait for the DB to initialize, wait to download entire blockchain, stop the node) and copy from that machine to the Pi Zero the `nxt_db` directory. 

5. Before running the node limit the Java VM memory usage to 700MB. I found this value that works well and also does not use much swap memory. You can fiddle with this value if your swap is very slow. 
  - Set this variable in the shell where the node will start, before starting it: `export mem_java="-Xms128M -Xmx700M"`

6. Run the node on the Pi Zero by executing `run.sh`.

### Please let me know in the comments if it worked for you or if you have issues. Any feedback is welcomed.

**Comment by 3MiR7** [comment]
Salut, având în vedere că e stoc 0 la tot ce înseamnă raspberry mă gândeam daca nu exista posibilitatea rulării de noduri pe telefoane android. Eu personal am 3 acasă funcționale dar cu ecranul spart. Ar putea reprezenta 3 noduri. 
Exista posibilitatea asta?

**Reply by georgeasiminei** [reply]
Salut @3MiR7, nu exista aceasta posibilitate si momentan nici nu avem planuri sa o implementam in viitor.

Translation for non-Romanian speakers to @3MiR7 question about running node on Android device:
There is no way to run the node on Android and currently no plan to do it.
